main/src/analytical_solutions/turbulence/density_pdf.cpp

```diff
@@ -0,0 +1,111 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <filesystem>
+#include <fstream>
+#include <numeric>
+
+#include "density_pdf.hpp"
+#include "io/arg_parser.hpp"
+#include "io/factory.hpp"
+#include "util/utils.hpp"
+#include "cstone/primitives/mpi_wrappers.hpp"
+
+using namespace sphexa;
+using T = float;
+
+int main(int argc, char** argv)
+{
+    auto [rank, numRanks] = initMpi();
+    const ArgParser parser(argc, (const char**)argv);
+
+    const std::string inputFile  = parser.get("--file");
+    const size_t      nBins      = parser.get("-n", 50);
+    std::string       outputFile = parser.get("-o", std::string("density_pdf.txt"));
+    const std::string sph_type   = parser.get("--sph", std::string("std"));
+    const T           minValue   = parser.get("--min", -8.0);
+    const T           maxValue   = parser.get("--max", 6.0);
+
+    if (!std::filesystem::exists(inputFile))
+    {
+        printf("Please provide a existing file: no file found at %s\n", inputFile.c_str());
+        return exitSuccess();
+    }
+
+    auto h5reader = fileReaderFactory(false, MPI_COMM_WORLD);
+    h5reader->setStep(inputFile, -1, FileMode::collective);
+
+    const size_t   localNumParticles  = h5reader->localNumParticles();
+    const size_t   globalNumParticles = h5reader->globalNumParticles();
+    std::vector<T> rho(localNumParticles);
+    std::vector<T> bins;
+
+    if (sph_type == "std") { h5reader->readField("rho", rho.data()); }
+    else
+    {
+        std::vector<T> m(localNumParticles);
+        std::vector<T> xm(localNumParticles);
+        h5reader->readField("kx", rho.data());
+        h5reader->readField("xm", xm.data());
+        h5reader->readField("m", m.data());
+#pragma omp for schedule(static)
+        for (size_t i = 0; i < localNumParticles; ++i)
+        {
+            rho[i] = rho[i] * m[i] / xm[i];
+        }
+    }
+    h5reader->closeStep();
+
+    T localTotalDensity = std::reduce(rho.begin(), rho.end());
+    T referenceDensity;
+    MPI_Allreduce(&localTotalDensity, &referenceDensity, 1, MpiType<T>{}, MPI_SUM, MPI_COMM_WORLD);
+    referenceDensity /= globalNumParticles;
+
+    bins = computeProbabilityDistribution(rho, referenceDensity, nBins, minValue, maxValue);
+    std::vector<T> reduced_bins(nBins, 0.0);
+    MPI_Reduce(bins.data(), reduced_bins.data(), nBins, MpiType<T>{}, MPI_SUM, 0, MPI_COMM_WORLD);
+
+    if (rank == 0)
+    {
+        std::for_each(reduced_bins.begin(), reduced_bins.end(),
+                      [globalNumParticles](T& i) { i /= globalNumParticles; });
```

Comment:
I think that the normalization factor should be the total number of data points * the width of the bin: globalNumParticles * binSize, because when you calculate the integral of the PDF you do sum(binvalue * binSize), so you have to divide by binSize in the normalization so that the area is 1.

Reply:
done!